server/src/main/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdSettings.java

```diff
@@ -207,6 +261,67 @@ public Iterator<Setting<?>> settings() {
 
     }
 
+    /**
+     * Validates that low, high and flood stage max headrooms adhere to the comparison: flood &lt; high &lt; low. Else, throws an exception.
+     */
+    static class MaxHeadroomValidator implements Setting.Validator<ByteSizeValue> {
+
+        @Override
+        public void validate(ByteSizeValue value) {
+
+        }
+
+        @Override
+        public void validate(final ByteSizeValue value, final Map<Setting<?>, Object> settings) {
+            final ByteSizeValue lowHeadroom = (ByteSizeValue) settings.get(CLUSTER_ROUTING_ALLOCATION_LOW_DISK_MAX_HEADROOM_SETTING);
+            final ByteSizeValue highHeadroom = (ByteSizeValue) settings.get(CLUSTER_ROUTING_ALLOCATION_HIGH_DISK_MAX_HEADROOM_SETTING);
+            final ByteSizeValue floodHeadroom = (ByteSizeValue) settings.get(
+                CLUSTER_ROUTING_ALLOCATION_DISK_FLOOD_STAGE_MAX_HEADROOM_SETTING
+            );
+
+            // For the comparisons, we need to mind that headroom values can default to -1.
+            if (floodHeadroom.compareTo(highHeadroom) > 0 && highHeadroom.getBytes() > 0) {
+                throw new IllegalArgumentException(
+                    "flood disk max headroom ["
+                        + floodHeadroom.getStringRep()
+                        + "] more than high disk max headroom ["
+                        + highHeadroom.getStringRep()
+                        + "]"
+                );
+            }
+            if (highHeadroom.compareTo(lowHeadroom) > 0 && lowHeadroom.getBytes() > 0) {
+                throw new IllegalArgumentException(
+                    "high disk max headroom ["
+                        + highHeadroom.getStringRep()
+                        + "] more than low disk max headroom ["
+                        + lowHeadroom.getStringRep()
+                        + "]"
+                );
+            }
+            // This check is also needed in case high headroom is -1 and relevant check above was skipped.
+            if (floodHeadroom.compareTo(lowHeadroom) > 0 && lowHeadroom.getBytes() > 0) {
```

Comment:
Hi @DaveCTurner , good catch! Indeed, this situation could arise:
For a disk with total 1000TB and the following settings:
low percentage: 85%, headroom 150GB 
high percentage: 90%, headroom -1
flood: 95%, headroom 20GB
Then the required minimum free disk space for the low watermark would be 100GB, for the high watermark it would be 10TB and for the flood watermark it would be 20GB. So indeed it could hit the high watermark before! I will add your check and also relevant tests if I can.

Reply:
Pushed commit with tests & new validation. Please check and unresolve conversation if it does not suffice. Thanks!